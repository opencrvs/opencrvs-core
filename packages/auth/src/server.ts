/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import * as Hapi from '@hapi/hapi'
import { DEFAULT_TIMEOUT } from '@auth/constants'
import { env } from '@auth/environment'
import authenticateHandler, {
  requestSchema as reqAuthSchema,
  responseSchema as resAuthSchema
} from '@auth/features/authenticate/handler'
import authenticateSuperUserHandler, {
  requestSchema as reqAuthSupSchema,
  responseSchema as resAuthSupSchema
} from '@auth/features/authenticateSuperUser/handler'
import verifyCodeHandler, {
  requestSchema as reqVerifySchema,
  responseSchma as resVerifySchema
} from '@auth/features/verifyCode/handler'
import refreshTokenHandler, {
  requestSchema as reqRefreshSchema,
  responseSchma as resRefreshSchema
} from '@auth/features/refresh/handler'
import resendNotificationHandler, {
  requestSchema as reqResendAuthenticationCodeSchema,
  responseSchma as resResendAuthenticationCodeSchema
} from '@auth/features/resend/handler'
import getPlugins from '@auth/config/plugins'
import * as database from '@auth/database'
import verifyTokenHandler, {
  reqVerifyTokenSchema,
  resVerifyTokenSchema
} from '@auth/features/verifyToken/handler'
import invalidateTokenHandler, {
  reqInvalidateTokenSchema
} from '@auth/features/invalidateToken/handler'
import verifyUserHandler, {
  requestSchema as reqVerifyUserSchema,
  responseSchema as resVerifyUserSchema
} from '@auth/features/retrievalSteps/verifyUser/handler'
import verifyNumberHandler, {
  requestSchema as reqVerifyNumberSchema,
  responseSchema as resVerifyNumberSchema
} from '@auth/features/retrievalSteps/verifyNumber/handler'
import verifySecurityQuestionHandler, {
  verifySecurityQuestionSchema,
  verifySecurityQuestionResSchema
} from '@auth/features/retrievalSteps/verifySecurityAnswer/handler'
import changePasswordHandler, {
  reqChangePasswordSchema
} from '@auth/features/retrievalSteps/changePassword/handler'
import sendUserNameHandler, {
  requestSchema as reqSendUserNameSchema
} from '@auth/features/retrievalSteps/sendUserName/handler'
import { tokenHandler } from '@auth/features/oauthToken/handler'
import { logger } from '@opencrvs/commons'
import { getPublicKey } from '@auth/features/authenticate/service'
import anonymousTokenHandler, {
  responseSchema
} from './features/anonymousToken/handler'
import { Boom, badRequest } from '@hapi/boom'

export type AuthServer = {
  server: Hapi.Server
  start: () => Promise<void>
  stop: () => Promise<void>
}

export async function createServer() {
  let whitelist: string[] = [env.DOMAIN]
  if (env.DOMAIN[0] !== '*') {
    whitelist = [env.COUNTRY_CONFIG_URL, env.LOGIN_URL, env.CLIENT_APP_URL]
  }
  logger.info(`Whitelist: ${JSON.stringify(whitelist)}`)
  const server = new Hapi.Server({
    host: env.AUTH_HOST,
    port: env.AUTH_PORT,
    routes: {
      cors: { origin: whitelist },
      payload: { maxBytes: 52428800, timeout: DEFAULT_TIMEOUT },
      response: {
        failAction: async (req, _2, err: Boom) => {
          if (process.env.NODE_ENV === 'production') {
            // In prod, log a limited error message and throw the default Bad Request error.
            logger.error(`Response validationError: ${err.message}`)
            throw badRequest(`Invalid response payload returned from handler`)
          } else {
            // During development, log and respond with the full error.
            logger.error(
              `${req.path} response has a validation error: ${err.message}`
            )
            throw err
          }
        }
      }
    }
  })

  /* add ping route by default for health check */
  server.route({
    method: 'GET',
    path: '/ping',
    handler: () =>
      // Perform any health checks and return true or false for success prop
      ({
        success: true
      }),
    options: {
      auth: false,
      tags: ['api'],
      description: 'Health check endpoint'
    }
  })
  server.route({
    method: 'GET',
    path: '/.well-known',
    handler: getPublicKey,
    options: {
      tags: ['api']
    }
  })

  // curl -H 'Content-Type: application/json' http://localhost:4040/anonymous-token
  server.route({
    method: 'GET',
    path: '/anonymous-token',
    handler: anonymousTokenHandler,
    options: {
      tags: ['api'],
      description: 'Authenticate an anonymous user',
      notes:
        'Returns a token to be used for endpoints that allow unauthorized access such as certificate verification endpoints',
      response: {
        schema: responseSchema
      }
    }
  })
  // curl -H 'Content-Type: application/json' -d '{"username": "test.user", "password": "test"}' http://localhost:4040/authenticate
  server.route({
    method: 'POST',
    path: '/authenticate',
    handler: authenticateHandler,
    options: {
      tags: ['api'],
      description: 'Authenticate with username and password',
      notes:
        'Authenticates user and returns nonce to use for collating the login for 2 factor authentication.' +
        'Sends an SMS to the user mobile with verification code',
      validate: {
        payload: reqAuthSchema
      },
      response: {
        schema: resAuthSchema
      }
    }
  })

  // curl -H 'Content-Type: application/json' -d '{"username": "test.user", "password": "test"}' http://localhost:4040/authenticate-super-user
  server.route({
    method: 'POST',
    path: '/authenticate-super-user',
    handler: authenticateSuperUserHandler,
    options: {
      tags: ['api'],
      description: 'Authenticate with username and password',
      notes:
        'Authenticates user and returns nonce to use for collating the login for 2 factor authentication.' +
        'Sends an SMS to the user mobile with verification code',
      validate: {
        payload: reqAuthSupSchema
      },
      response: {
        schema: resAuthSupSchema
      }
    }
  })

  // curl -H 'Content-Type: application/json' -d '{"nonce": ""}' http://localhost:4040/resendAuthenticationCode
  server.route({
    method: 'POST',
    path: '/resendAuthenticationCode',
    handler: resendNotificationHandler,
    options: {
      tags: ['api'],
      description: 'Resend another authentication code',
      notes:
        'Sends a new authentication code to the user based on the phone number or email associated with the nonce',
      validate: {
        payload: reqResendAuthenticationCodeSchema
      },
      response: {
        schema: resResendAuthenticationCodeSchema
      }
    }
  })

  // curl -H 'Content-Type: application/json' -d '{"code": "123456"}' http://localhost:4040/verifyCode
  server.route({
    method: 'POST',
    path: '/verifyCode',
    handler: verifyCodeHandler,
    options: {
      tags: ['api'],
      description: 'Verify the 2 factor auth code',
      notes:
        'Verifies the 2 factor auth code and returns the JWT API token for future requests',
      validate: {
        payload: reqVerifySchema
      },
      response: {
        schema: resVerifySchema
      }
    }
  })

  // curl -H 'Content-Type: application/json' -d '{"nonce": "", "token": ""}' http://localhost:4040/refreshToken
  server.route({
    method: 'POST',
    path: '/refreshToken',
    handler: refreshTokenHandler,
    options: {
      tags: ['api'],
      description: 'Refresh an expiring token',
      notes:
        'Verifies the expired client token as true and returns a refreshed JWT API token for future requests',
      validate: {
        payload: reqRefreshSchema
      },
      response: {
        schema: resRefreshSchema
      }
    }
  })

  // curl -H 'Content-Type: application/json' -d '{ "token": "" }' http://localhost:4040/verifyToken
  server.route({
    method: 'POST',
    path: '/verifyToken',
    handler: verifyTokenHandler,
    options: {
      tags: ['api'],
      description: 'Check if a token is marked as invalid or not',
      notes:
        'Check if this token is part of the invalid token list that is stored in redis',
      validate: {
        payload: reqVerifyTokenSchema
      },
      response: {
        schema: resVerifyTokenSchema
      }
    }
  })

  // curl -H 'Content-Type: application/json' -d '{ "token": "" }' http://localhost:4040/invalidateToken
  server.route({
    method: 'POST',
    path: '/invalidateToken',
    handler: invalidateTokenHandler,
    options: {
      tags: ['api'],
      description: 'Marks token as invalid until it expires',
      notes:
        'Adds a token to the invalid tokens stored in Redis, ' +
        'these are stored as individual key value pairs to that we can set their expiry TTL individually',
      validate: {
        payload: reqInvalidateTokenSchema
      },
      response: {
        schema: false
      }
    }
  })

  // curl -H 'Content-Type: application/json' -d '{ "mobile": "" }' http://localhost:4040/verifyUser
  server.route({
    method: 'POST',
    path: '/verifyUser',
    handler: verifyUserHandler,
    options: {
      tags: ['api'],
      description:
        'First step of password or username retrieval steps.' +
        'Check if user exists for given mobile number or not.',
      notes:
        'Verifies user and returns nonce to use for next step of password reset flow.' +
        'Sends an SMS to the user mobile with verification code',
      validate: {
        payload: reqVerifyUserSchema
      },
      response: {
        schema: resVerifyUserSchema
      }
    }
  })

  // curl -H 'Content-Type: application/json' -d '{ "mobile": "" }' http://localhost:4040/verifyUser

  server.route({
    method: 'POST',
    path: '/verifyNumber',
    handler: verifyNumberHandler,
    options: {
      tags: ['api'],
      description:
        'Second step of password or username retrieval steps.' +
        'Check if provided verification code is valid or not.',
      notes:
        'Verifies code for given nonce and returns a random security question for that user.',
      validate: {
        payload: reqVerifyNumberSchema
      },
      response: {
        schema: resVerifyNumberSchema
      }
    }
  })

  // curl -H 'Content-Type: application/json' -d '{ "questionKey": "", "answer": "", "nonce": "" }' http://localhost:4040/verifyUser
  server.route({
    method: 'POST',
    path: '/verifySecurityAnswer',
    handler: verifySecurityQuestionHandler,
    options: {
      tags: ['api'],
      description:
        'Third step of password or username retrieval steps.' +
        'Checks if the submitted security question answer is right',
      notes:
        'Verifies security answer and updates the nonce information so that it can be used for changing the password' +
        'In-case of a wrong answer, it will return an another question key.',
      validate: {
        payload: verifySecurityQuestionSchema
      },
      response: {
        schema: verifySecurityQuestionResSchema
      }
    }
  })

  // curl -H 'Content-Type: application/json' -d '{ "newPassword": "", "nonce": "" }' http://localhost:4040/changePassword
  server.route({
    method: 'POST',
    path: '/changePassword',
    handler: changePasswordHandler,
    options: {
      tags: ['api'],
      description:
        'Final step of password retrieval flow.' + 'Changes the user password',
      notes:
        'Expects the nonce parameter to be coming from the reset password journey',
      validate: {
        payload: reqChangePasswordSchema
      },
      response: {
        schema: false
      }
    }
  })

  // curl -H 'Content-Type: application/json' -d '{ "nonce": "" }' http://localhost:4040/sendUserName
  server.route({
    method: 'POST',
    path: '/sendUserName',
    handler: sendUserNameHandler,
    options: {
      tags: ['api'],
      description:
        'Final step of username retrieval flow.' +
        'Sends the username to user mobile number',
      notes:
        'Expects the nonce parameter to be coming from the retrieve username journey',
      validate: {
        payload: reqSendUserNameSchema
      },
      response: {
        schema: false
      }
    }
  })

  server.route({
    method: 'POST',
    path: '/token',
    handler: tokenHandler,
    options: {
      tags: ['api'],
      description: 'Authenticate system with client_id and client_secret'
    }
  })

  await server.register(getPlugins())
  server.ext({
    type: 'onRequest',
    method(
      request: Hapi.Request & {
        sentryScope?: {
          setExtra: (key: string, value: unknown) => void
        }
      },
      h
    ) {
      request.sentryScope?.setExtra('payload', request.payload)
      return h.continue
    }
  })

  async function stop() {
    await server.stop()
    await database.stop()
    server.log('info', 'server stopped')
  }

  async function start() {
    await server.start()
    await database.start()
    server.log('info', `server started on ${env.AUTH_HOST}:${env.AUTH_PORT}`)
  }

  return { server, start, stop }
}
