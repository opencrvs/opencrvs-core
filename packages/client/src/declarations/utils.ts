/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import { EVENT_STATUS } from '@client/workqueue'
import { SUBMISSION_STATUS } from '.'

export const isPendingCorrection = (status: SUBMISSION_STATUS | undefined) =>
  status === EVENT_STATUS.CORRECTION_REQUESTED

export const isReviewableDeclaration = (
  status: SUBMISSION_STATUS | undefined
) => status && [EVENT_STATUS.DECLARED, EVENT_STATUS.VALIDATED].includes(status)

export const isUpdatableDeclaration = (status: SUBMISSION_STATUS | undefined) =>
  status &&
  [
    SUBMISSION_STATUS.DRAFT,
    EVENT_STATUS.IN_PROGRESS,
    EVENT_STATUS.REJECTED
  ].includes(status)

export const isPrintable = (status: SUBMISSION_STATUS | undefined) =>
  status &&
  [SUBMISSION_STATUS.REGISTERED, SUBMISSION_STATUS.ISSUED].includes(status)

export const isIssuable = (status: SUBMISSION_STATUS | undefined) =>
  status === SUBMISSION_STATUS.CERTIFIED

export const isRecordOrDeclaration = (status: SUBMISSION_STATUS | undefined) =>
  status
    ? [
        SUBMISSION_STATUS.REGISTERED,
        SUBMISSION_STATUS.CORRECTION_REQUESTED,
        SUBMISSION_STATUS.CERTIFIED
      ].includes(status)
      ? 'record'
      : 'declaration'
    : ''

export const canBeCorrected = (status: SUBMISSION_STATUS | undefined) =>
  status &&
  [
    SUBMISSION_STATUS.REGISTERED,
    SUBMISSION_STATUS.CERTIFIED,
    SUBMISSION_STATUS.ISSUED
  ].includes(status)

export const isArchivable = (status: SUBMISSION_STATUS | undefined) =>
  status &&
  [
    SUBMISSION_STATUS.IN_PROGRESS,
    SUBMISSION_STATUS.DECLARED,
    SUBMISSION_STATUS.VALIDATED,
    SUBMISSION_STATUS.REJECTED
  ].includes(status)

export const canBeReinstated = (status: SUBMISSION_STATUS | undefined) =>
  status === SUBMISSION_STATUS.ARCHIVED

export const isViewable = (status: SUBMISSION_STATUS | undefined) =>
  status !== SUBMISSION_STATUS.DRAFT
