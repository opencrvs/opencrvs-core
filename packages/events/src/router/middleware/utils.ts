/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import type { AnyTRPCMiddlewareFunction } from '@trpc/server'
import {
  ActionInputWithType,
  TokenWithBearer,
  TokenUserType
} from '@opencrvs/commons'

export interface UserDetails {
  type: TokenUserType.USER | TokenUserType.SYSTEM
  id: string
  primaryOfficeId: string | undefined
  role: string
}

export type Context = { token: TokenWithBearer; user: UserDetails }
/**
 * TRPC Middleware options with correct context.
 * Actual middleware type definition is only for internal use within TRPC.
 */
export type MiddlewareOptions = Omit<
  Parameters<AnyTRPCMiddlewareFunction>[0],
  'ctx'
> & { ctx: Context }

export type ActionMiddlewareOptions = Omit<MiddlewareOptions, 'input'> & {
  input: ActionInputWithType
}
