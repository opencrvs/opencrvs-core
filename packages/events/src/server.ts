/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import { createServer, IncomingMessage } from 'http'
import { createOpenApiHttpHandler } from 'trpc-to-openapi'
import { createHTTPHandler } from '@trpc/server/adapters/standalone'
import '@opencrvs/commons/monitoring'
import { logger } from '@opencrvs/commons'
import { appRouter } from './router/router'
import { createContext } from './context'

function stringifyRequest(req: IncomingMessage) {
  const url = new URL(req.url || '', `http://${req.headers.host}`)
  return `'${req.method} ${url.pathname}'`
}

const trpcConfig: Parameters<typeof createHTTPHandler>[0] = {
  router: appRouter,
  middleware: (req, _, next) => {
    logger.info(`Request: ${stringifyRequest(req)}`)
    return next()
  },
  onError: ({ error, req }) => {
    logger.warn(
      `Error for request: ${stringifyRequest(req)}. Error: '${error.message}'`
    )
  },
  createContext
}

function isTrpcRequest(req: IncomingMessage): boolean {
  if (!req.url) {
    throw new Error('No URL provided')
  }
  const url = new URL(req.url, `http://${req.headers.host}`)
  const pathName = url.pathname.replace(/^\//, '')
  const trpcProcedurePaths = Object.keys(appRouter._def.procedures)

  return (
    url.search.startsWith('?input') || trpcProcedurePaths.includes(pathName)
  )
}

export function server() {
  const restServer = createOpenApiHttpHandler(trpcConfig)
  const trpcServer = createHTTPHandler(trpcConfig)

  return createServer((req, res) => {
    if (!req.url) {
      res.writeHead(500)
      res.end('No URL provided')
      return
    }

    // If it's a tRPC request, handle it with the tRPC server
    if (isTrpcRequest(req)) {
      trpcServer(req, res)
    } else {
      // If it's a REST request, handle it with the REST server
      void restServer(req, res)
    }
  })
}
