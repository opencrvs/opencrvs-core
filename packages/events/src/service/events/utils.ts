/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import { Action, ActionType } from '@opencrvs/commons'

export function findLastAssignmentAction(actions: Action[]) {
  return actions
    .filter(
      ({ type }) => type === ActionType.ASSIGN || type === ActionType.UNASSIGN
    )
    .reduce<
      Action | undefined
    >((latestAction, action) => (!latestAction || action.createdAt > latestAction.createdAt ? action : latestAction), undefined)
}
