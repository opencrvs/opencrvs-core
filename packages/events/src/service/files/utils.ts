/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */
import { differenceWith, first, isEqual, orderBy } from 'lodash'
import {
  Draft,
  EventConfig,
  EventDocument,
  FieldConfig,
  FileFieldType,
  getActionAnnotationFields,
  AnyFileField,
  getDeclarationFields,
  getMixedPath,
  FileFieldWithOptionValue,
  FileFieldValue,
  ActionUpdate,
  getCurrentEventStateWithDrafts,
  EventState,
  getCurrentEventState,
  ActionDocument
} from '@opencrvs/commons'

function isFieldWithFile(field: FieldConfig): field is AnyFileField {
  return FileFieldType.safeParse(field.type).success
}

/**
 *
 * @param field File field
 * @param update single action or event state.
 * @returns normalised file reference based on the field id and event state.
 */
function getFileReferences(
  field: AnyFileField,
  update: ActionUpdate | EventState
): { id: string; value: string }[] {
  const value = getMixedPath<AnyFileField['defaultValue']>(update, field.id)

  if (!value) {
    return []
  }

  const maybeFileOptions = FileFieldWithOptionValue.safeParse(value)
  if (maybeFileOptions.success) {
    return maybeFileOptions.data.map((opt) => ({
      id: field.id,
      value: opt.path
    }))
  }

  const maybeFile = FileFieldValue.safeParse(value)

  if (maybeFile.success) {
    return [{ id: field.id, value: maybeFile.data.path }]
  }

  if (typeof value === 'string') {
    return [{ id: field.id, value }]
  }

  throw new Error(
    `Unknown filetype, ${field.type} with value: ${JSON.stringify(value)}`
  )
}

/**
 *  Since API allows sending PATCH requests with only the changed fields,
 *  we need to check the current event state with the current draft.
 * @returns file references that are not present in the current event state but are when previous draft is applied to the state.
 */
function getUnreferencedDeclarationDraftFiles({
  event,
  currentDraft,
  previousDraft,
  configuration
}: {
  event: EventDocument
  currentDraft?: Draft
  previousDraft: Draft
  configuration: EventConfig
}) {
  const currentEventState = currentDraft
    ? getCurrentEventStateWithDrafts({
        event,
        draft: currentDraft,
        configuration
      })
    : getCurrentEventState(event, configuration)

  const previousEventState = getCurrentEventStateWithDrafts({
    event,

    draft: {
      // Given previous draft, if event has been updated afterwards (running deletion after submission of actual action.) the data would never get applied.
      ...previousDraft,
      createdAt: new Date().toISOString(),
      action: {
        ...previousDraft.action,
        createdAt: new Date().toISOString()
      }
    },
    configuration
  })

  const declarationFileFields =
    getDeclarationFields(configuration).filter(isFieldWithFile)

  const currentDeclarationFiles = declarationFileFields.flatMap((field) =>
    getFileReferences(field, currentEventState.declaration)
  )

  const previousDeclarationFiles = declarationFileFields.flatMap((field) =>
    getFileReferences(field, previousEventState.declaration)
  )

  return differenceWith(
    previousDeclarationFiles,
    currentDeclarationFiles,
    isEqual
  )
}

function getUnreferencedAnnotationDraftFiles({
  currentAction,
  previousDraft,
  configuration
}: {
  currentAction: Draft['action'] | ActionDocument
  previousDraft: Draft
  configuration: EventConfig
}) {
  if (currentAction.type !== previousDraft.action.type) {
    // eslint-disable-next-line no-console
    console.warn(
      'Previous draft was for different action. Not comparing file references between them.'
    )
    return []
  }

  const currentDraftActionConfig = configuration.actions.find(
    (action) => action.type === previousDraft.action.type
  )

  if (!currentDraftActionConfig) {
    // eslint-disable-next-line no-console
    console.warn(
      `No configuration found for action ${previousDraft.action.type}.`
    )
  }

  const annotationFileFields = currentDraftActionConfig
    ? getActionAnnotationFields(currentDraftActionConfig).filter(
        isFieldWithFile
      )
    : []

  // @TODO: There might be a case where annotation shares id with declaration field? Should it limited by the configuration?
  const currentAnnotationFiles = annotationFileFields.flatMap((field) =>
    getFileReferences(field, currentAction.annotation ?? {})
  )

  const previousAnnotationFiles = annotationFileFields.flatMap((field) =>
    getFileReferences(field, previousDraft.action.annotation ?? {})
  )

  return differenceWith(
    previousAnnotationFiles,
    currentAnnotationFiles,
    isEqual
  )
}

/**
 *
 * @returns all the filenames that are referenced in the current draft, but not in the previous draft.
 */
export function getUnreferencedDraftFiles({
  event,
  currentDraft,
  previousDraft,
  configuration
}: {
  event: EventDocument
  /**
   * Current draft is optional, because we might be comparing the previous draft with the current event state. (e.g. actual action has been submitted)
   */
  currentDraft?: Draft
  previousDraft: Draft
  configuration: EventConfig
}) {
  const unreferencedDeclarationDraftFiles =
    getUnreferencedDeclarationDraftFiles({
      event,
      currentDraft,
      previousDraft,
      configuration
    })

  const currentAction =
    currentDraft?.action ??
    first(
      orderBy(
        event.actions.filter(
          (action) => action.type === previousDraft.action.type
        ),
        ['createdAt'],
        'desc'
      )
    )

  if (!currentAction) {
    // This should not happen. It means that the previous draft action type was invalid.
    // eslint-disable-next-line no-console
    console.error(
      `Could not find corresponding action ${previousDraft.action.type} for event: ${event.id}`
    )

    return unreferencedDeclarationDraftFiles
  }

  const unreferencedAnnotationDraftFiles = getUnreferencedAnnotationDraftFiles({
    // @TODO: for some reason declaration is omitted from some type?
    currentAction: currentAction as Draft['action'] | ActionDocument,
    previousDraft,
    configuration
  })

  return [
    ...unreferencedDeclarationDraftFiles,
    ...unreferencedAnnotationDraftFiles
  ]
}
