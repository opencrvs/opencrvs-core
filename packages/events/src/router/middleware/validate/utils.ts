/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import { TRPCError } from '@trpc/server'
import _ from 'lodash'
import {
  ActionUpdate,
  FieldConfig,
  getFieldValidationErrors,
  Inferred,
  errorMessages
} from '@opencrvs/commons/events'

type ValidationError = {
  message: string
  id: string
  value: unknown
}

export function getFormFieldErrors(formFields: Inferred[], data: ActionUpdate) {
  return formFields.reduce(
    (errorResults: ValidationError[], field: FieldConfig) => {
      const fieldErrors = getFieldValidationErrors({
        field,
        values: data
      }).errors

      if (fieldErrors.length === 0) {
        return errorResults
      }

      // For backend, use the default message without translations.
      const errormessageWithId = fieldErrors.map((error) => ({
        message: error.message.defaultMessage,
        id: field.id,
        value: data[field.id as keyof typeof data]
      }))

      return [...errorResults, ...errormessageWithId]
    },
    []
  )
}

export function getVerificationPageErrors(
  verificationPageIds: string[],
  data: ActionUpdate
) {
  return verificationPageIds
    .map((pageId) => {
      const value = data[pageId]
      return typeof value !== 'boolean'
        ? {
            message: 'Verification page result is required',
            id: pageId,
            value
          }
        : null
    })
    .filter((error) => error !== null)
}

export function throwWhenNotEmpty(errors: unknown[]) {
  if (errors.length > 0) {
    throw new TRPCError({
      code: 'BAD_REQUEST',
      message: JSON.stringify(errors)
    })
  }
}

function isObject(value: unknown): value is Record<string, unknown> {
  return _.isPlainObject(value)
}

/**
 * Flattens an object into a list of [key, value] pairs using dot/bracket notation.
 */
function flattenEntries<T>(obj: T, path: string = ''): [string, unknown][] {
  if (Array.isArray(obj)) {
    return obj.flatMap((value, index) =>
      flattenEntries(value, `${path}[${index}]`)
    )
  }

  if (isObject(obj)) {
    return Object.entries(obj).flatMap(([key, value]) =>
      flattenEntries(value, path ? `${path}.${key}` : key)
    )
  }

  return obj === undefined ? [] : [[path, obj]]
}

/**
 * Compares update vs cleaned payloads and returns an array of offending keys.
 */
export function getInvalidUpdateKeys<T>({
  update,
  cleaned
}: {
  update: T
  cleaned: T
}): ValidationError[] {
  const updateEntries = flattenEntries(update)
  const cleanedKeys = flattenEntries(cleaned).map(([key]) => key)

  return updateEntries
    .filter(([key]) => !cleanedKeys.includes(key))
    .map(([key, value]) => ({
      message: errorMessages.hiddenField.defaultMessage,
      id: key,
      value
    }))
}
