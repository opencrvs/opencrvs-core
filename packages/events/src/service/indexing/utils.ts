/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */
import { EventSearchIndex } from '@opencrvs/commons/events'
import {
  QueryDslQueryContainer,
  SearchRequest
} from '@elastic/elasticsearch/lib/api/types'

const FIELD_SEPARATOR = '____'
export const DEFAULT_SIZE = 10

/**
 * Generates an Elasticsearch query to search within `document.data`
 * using the provided search payload.
 */
export function generateQuery(event: Omit<EventSearchIndex, 'type'>) {
  const must: QueryDslQueryContainer[] = Object.entries(event).map(
    ([key, value]) => ({
      match: {
        [`data.${key.replaceAll('.', FIELD_SEPARATOR)}`]: value
      }
    })
  )

  return {
    bool: {
      must
    }
  } as SearchRequest['query']
}
