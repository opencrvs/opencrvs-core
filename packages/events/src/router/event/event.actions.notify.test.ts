/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import { createTestClient, setupTestCase } from '@events/tests/utils'
import { SCOPES } from '@opencrvs/commons'
import { TRPCError } from '@trpc/server'

test(`prevents forbidden access if missing required scope`, async () => {
  const { user } = await setupTestCase()
  const client = createTestClient(user, [])

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  await expect(client.event.actions.notify({} as any)).rejects.toMatchObject(
    new TRPCError({ code: 'FORBIDDEN' })
  )
})

test(`allows access if required scope is present`, async () => {
  const { user } = await setupTestCase()
  const client = createTestClient(user, [SCOPES.RECORD_SUBMIT_INCOMPLETE])

  await expect(
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    client.event.actions.notify({} as any)
  ).rejects.not.toMatchObject(new TRPCError({ code: 'FORBIDDEN' }))
})
