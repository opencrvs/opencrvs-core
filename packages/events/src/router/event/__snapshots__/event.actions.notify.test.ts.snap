// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`event.actions.notify > allows sending partial payload as NOTIFY action 1`] = `
{
  "applicant.address": {
    "addressType": "DOMESTIC",
    "country": "FAR",
    "district": "5ef450bc-712d-48ad-93f3-8da0fa453baa",
    "number": "55",
    "province": "a45b982a-5c7b-4bd9-8fd8-a42d0994054c",
    "residentialArea": "Example Residential Area",
    "street": "Example Street",
    "town": "Example Town",
    "urbanOrRural": "URBAN",
    "zipCode": "123456",
  },
  "applicant.age": 19,
  "applicant.dobUnknown": true,
  "applicant.image": {
    "filename": "4f095fc4-4312-4de2-aa38-86dcc0f71044.png",
    "originalFilename": "abcd.png",
    "type": "image/png",
  },
  "recommender.none": true,
}
`;
